[DOC] PyMC, Aesara and  Aeppl intro notebook

juanitorduz opened the pull request:
This PR partially solves https://github.com/pymc-devs/pymc/issues/5538. 

I am trying to add comments and understand @ricardoV94's nice intro notebook. Nevertheless, I still have a long way to fully understand. Specifically, starting from section `Graph manipulation 102` things become less clear for me. Still, as suggested I am opening this PR to get feedback and iterate.

Commit:
add raw version of the notebook

Commit:
first section on aesara

Commit:
random variables section

Commit:
update

Commit:
Merge branch 'main' into docs/aesara_pymc_nb

juanitorduz commented:
@ricardoV94 What do you thing should be the scope of this notebook? Specifically, which sections we should focus on and which ones we should remove (or postpone)?

ricardoV94 commented:
> @ricardoV94 What do you thing should be the scope of this notebook? Specifically, which sections we should focus on and which ones we should remove (or postpone)?

For now you can focus on PyMC returning RandomVariables (you can skip most of the Aesara Op and specially the manual seeding bit), and the "vanilla" logprob part (you can skip the Aeppl bits as well). What's left would be the most critical part

Commit:
prune notebook

juanitorduz commented:
@ricardoV94 in https://github.com/pymc-devs/pymc/pull/5721/commits/54b2e69b1ccad8338efb4e757bf90a90445c4470 I removed the parts of the notebook which are not needed based in your input. Would you agree that the current content is what I need to develop? I think I would have some time these upcoming weeks to work on this notebook (at least some times during the evenings)

Two small remarks on code errors (which you can see on the notebook):
- In my current version (`4.b06`) I am getting `AttributeError: 'Model' object has no attribute 'compute_initial_point'`
- At the beginning of the section *What is the deal with those value variables in the model?* I whet an error because `rv` is not defined, what should `rv` be?

Thanks 🙏 !

OriolAbril commented:
> In my current version (4.b06)

I can't speak about the method in question, but you should not be executing that notebook on a release and instead do that on the development version aka local aka from github main. This notebook will be part of the _versioned_ docs and will be re-executed every time the docs are built (so it can't take very long to run either). 

I'll try to review the formatting, style and cross-references of the notebook some time next week.

juanitorduz commented:
@OriolAbril so this means I should use the local development environment described in https://github.com/pymc-devs/pymc/wiki/Installation-Guide-(MacOS)#pymc-v4-installation-on-macos ? This way I am installing from the `main` branch?

Commit:
add intro section and simple examples

OriolAbril commented:
That looks right. Just to be sure, you'll find better guidance and avise on the contributing section of the documentation (when reading contributing docs always do so on the "latest" version/tag): https://docs.pymc.io/en/latest/contributing/index.html.

Commit:
rename nb and add section on intro aesara graphs

ricardoV94 commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1767 @@
+{
```
<p>That edge case is a great proof but I worry it may concern users/ start less useful discussions in relation to this notebook goals</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='27'/>

juanitorduz replied:
<p>mmm Ok. I do like it as a "proof" of the graph simplification. But I can of course remove it :)</p>

twiecki replied:
<p>specially --&gt; specify</p>

twiecki replied:
<p>But yeah, I don't think it adds much, agreeing with Ricardo.</p>

juanitorduz replied:
<p>Alright! I will remove it :) </p>

ricardoV94 commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1767 @@
+{
```
<p>Let's not use JAX, that was used in the presentation because it's a new thing for devs, but might just confuse/distract users </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='49'/>

juanitorduz replied:
<p>Sure!</p>

ricardoV94 commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1767 @@
+{
```
<p>We should explain the transformed value variables. I think the original doc we are removing also did that. May be nice to first show a model without transformed variables and then one with them. I emphasize this because it's a common source of confusion </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='95'/>

juanitorduz replied:
Sure! I will keep this one in mind once I get to this section :)

ricardoV94 commented:
> Would you agree that the current content is what I need to develop? 

Yes it looks about right. And I love the changes you already added!

> Two small remarks on code errors (which you can see on the notebook):
> - In my current version (`4.b06`) I am getting `AttributeError: 'Model' object has no attribute 'compute_initial_point'`

You might have already figured it out, but we renamed it to `initial_point` to be more succint.

> At the beginning of the section *What is the deal with those value variables in the model?* I whet an error because `rv` is not defined, what should `rv` be?

I'll have to check but I probably reordered some part and didn't fix it. It was a scipy frozen rv like 

```python
import scipy.stats
rv = scipy.stats.norm(mu, sigma)
```
With whatever values for mu and sigma make sense

Commit:
add random samples section

Commit:
improve notation

twiecki commented:
Can we make an abridged version here that we can merge and then improve it in subsequent PRs?

Commit:
clean notebook and extract main sections [WIP]

juanitorduz commented:
@ricardoV94 Sorry about coming back to this notebook so late but I have been quite busy 😓 ! Today I revisited it and cleaned and extracted the most relevant pars of the initial notebook. I could see myself  working in this notebooks for days but in view of @twiecki 's comment about having an initial version to iterate with my questions are:

- What are the essentials parts to be added or removed?
- Any feedback for the first iteration?

As you have a better overview of the documentation, your input about how to make this notebook useful and not redundant would be great! Based on you input I will work to ship this one (at least the first iteration) soon 🙏 .

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>Should we add this to examples instead as a deep-dive?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

ricardoV94 replied:
<p>This will replace our old developer guide, so I think it belongs in the core documentation</p>

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>Maybe explain more what can be seen here. Like Elemwise, composite? InplaceDimShuffle? Just to give some intuition.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='14'/>

ricardoV94 replied:
<p>Or perhaps not dprint the function, the next example where we print the function does not introduce new <code>Op</code>s  <span class="ql-cursor">﻿</span></p>

juanitorduz replied:
<p>I will remove the <code>dprint</code>.</p>

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>I would add a sentence like: "The actual code is not as important here, the focus is on the outputs."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='33'/>

juanitorduz replied:
<p>+1</p>

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>manipulate the computational graph, **something that is not possible with TensorFlow or PyTorch**.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='37'/>

juanitorduz replied:
<p>+1</p>

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>unnecessary composition **after we eval()'ed it**.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

ricardoV94 replied:
<p>What do you mean?</p>

juanitorduz replied:
<p>I also do not understand the suggestion :)</p>

twiecki replied:
I meant to add something to the end of the sentence that says that it removes the part from the graph only after we eval'ed it. Because a user might have expected it to be removed the first time you print it, but only by eval-ing it does it apply the simplification. Super minor.

juanitorduz replied:
Ok! I added this to the remark. Is it better now?

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>This is nothing else *than*?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='79'/>

juanitorduz replied:
<p>yes! Thx!</p>

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>I'd include an example that actually works. Otherwise I wouldn't even expect it to work because it's not a RV directly.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='85'/>

juanitorduz replied:
<p>Any particular suggestion on an interesting example?</p>

twiecki replied:
I'm actually surprised it doesn't work with `cumsum()`.  @ricardoV94 would know.

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>If these haven't been introduced before it's a bit odd to ask what the deal is with something I haven't heard before. Maybe "What are value variables and why are they important?"</p><p> </p><p>And then I would add explanation as to what it is.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='87'/>

ricardoV94 replied:
<p>Agree</p>

juanitorduz replied:
@ricardoV94  Maybe you can help me with a few sentences for this intro! You have a much more in-depth knowledge than me to choose the right words :)

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>Maybe split out over 3 cells? The () is a bit odd to parse.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='88'/>

juanitorduz replied:
<p>yep!</p>

twiecki commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>Add text cell above setting up what you're showing. "Let's look at how these value variables behave in a simple model".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='89'/>

juanitorduz replied:
<p>+1</p>

twiecki commented:
Minor comments but this looks great!

ricardoV94 commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>Maybe it's worth to mention what the inputs are, as this is an important part of the RandomVariable API. I am not totally convinced though...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='56'/>

juanitorduz replied:
<p>By "inputs" you mean <code>RandomGeneratorSharedVariable</code> and / or what else? </p>

ricardoV94 replied:
<p>Yes</p>

juanitorduz replied:
Added some comments from the original notebook in https://github.com/pymc-devs/pymc/pull/5721/commits/b0781cf835d30257a8458bac80c76db3556936a6

juanitorduz replied:
<p>Added some comments from the original notebook in https://github.com/pymc-devs/pymc/pull/5721/commits/b0781cf835d30257a8458bac80c76db3556936a6</p>

ricardoV94 commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1819 @@
+{
```
<p>Actually this example will now work on main xD... Let's just not include it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='84'/>

juanitorduz replied:
<p>hahah just updated the dev environment with <code>main</code> and you are right! I'll remove it! </p>

ricardoV94 commented:
@juanitorduz this looks awesome. We can definitely merge after you address the comments that were left in the reviews, and then we can consider expanding or iterating in future PRs if you have ideas about that.

Commit:
Merge branch 'main' into docs/aesara_pymc_nb

Commit:
fix suggestions part 1

Commit:
fix suggestions part 2

juanitorduz commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1783 @@
+{
```
<p>@<a href="https://github.com/ricardoV94" target="_blank">ricardoV94</a> Maybe you can help me with a few sentences for this intro! You have a much more in-depth knowledge than me to choose the right words :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='82'/>

ricardoV94 replied:
<p>What about something along these lines?</p><ol><li>Logp graph does not have random variables (obviously)</li><li>Instead it's defined in terms of input (value) variables</li><li>Each RV is replaced by a respective input (value) variable</li></ol><p> </p>

juanitorduz replied:
Added these comments in https://github.com/pymc-devs/pymc/pull/5721/commits/b0781cf835d30257a8458bac80c76db3556936a6

juanitorduz replied:
<p>Added these comments in https://github.com/pymc-devs/pymc/pull/5721/commits/b0781cf835d30257a8458bac80c76db3556936a6</p>

Commit:
fix suggestions part 3

ricardoV94 commented on docs/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1783 @@
+{
```
<p>Maybe simplify to:</p><p> </p><p><strong>Remark:</strong> Again, note that <code>aesara</code> is clever enough to omit the <code>exp</code> and <code>log</code> once we compile the function.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='46'/>

juanitorduz replied:
<p>+1</p>

juanitorduz replied:
+1

OriolAbril reviewed (commented):
This notebook should be in the core notebooks folder: https://github.com/pymc-devs/pymc/tree/main/docs/source/learn/core_notebooks and listed in the toctree of https://github.com/pymc-devs/pymc/blob/main/docs/source/learn/core_notebooks/index.md

Commit:
fix suggestions part 3

Commit:
move notebook to right folder

juanitorduz commented:
> c1dcc89

Thanks! Fixed this in https://github.com/pymc-devs/pymc/pull/5721/commits/c1dcc8936b11b07b0f3b5607d50a9dd26d3d2e2c

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1893 @@
+{
```
<p>There should be a myst target at the beginning of the file: <code>(pymc_aesara)=</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

juanitorduz replied:
Thanks! I added it in https://github.com/pymc-devs/pymc/pull/5721/commits/54841a3fd0c1b57714d0df6504fcc4d342220933

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1893 @@
+{
```
<p>*One</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='21'/>

juanitorduz replied:
Thanks! Solved in https://github.com/pymc-devs/pymc/pull/5721/commits/54841a3fd0c1b57714d0df6504fcc4d342220933

Commit:
remove links

OriolAbril commented:
Output of the failing pre-commit check:
* ``docs/source/learn/core_notebooks/pymc_aesara.ipynb:13:        "In this notebook we want to give an introduction of how PyMC models translate to Aesara graphs. The purpose is not to give a detailed description of all [`aesara`](https://github.com/aesara-devs/aesara)'s capabilities but rather focus on the main concepts to understand its connection with PyMC. For a more detailed description of the project please refer to the [official documentation](https://aesara.readthedocs.io/en/latest/index.html).\n",``
  * Use `` {doc}`official documentation <aesara:index>` `` instead
* ``docs/source/learn/core_notebooks/pymc_aesara.ipynb:178:        "We can use the [`aesara.dprint`](https://aesara.readthedocs.io/en/latest/library/index.html#aesara.dprint) function to print the computational graph of any given tensor."``
  * Use `` {func}`aesara.dprint` ``
* ``docs/source/learn/core_notebooks/pymc_aesara.ipynb:222:        "Note that this graph does not do any computation (yet!). It is simply defining the sequence of steps to be done. The aesara function [`aesara.function`](https://aesara.readthedocs.io/en/latest/library/compile/function.html#aesara.compile.function.function) is used to define a callable object so that we can push values trough the graph."``
  * Use `` {func}`~aesara.compile.function.function` `` will only show `function` or `` {func}`aesara.function <aesara.compile.function.function>` `` to show `aesara.function`. It looks like aesara still documents objects at a different import path from the one they are used from.
* ``docs/source/learn/core_notebooks/pymc_aesara.ipynb:586:        "Note that this is very similar to the output of the  [`aesara.dprint`](https://aesara.readthedocs.io/en/latest/library/index.html#aesara.dprint) function introduced above."``
  * Repeated
* ``docs/source/learn/core_notebooks/pymc_aesara.ipynb:724:        "To modify the graph we need to use the [`aesara.clone_replace`](https://aesara.readthedocs.io/en/latest/library/index.html?highlight=clone_replace#aesara.clone_replace) function, which *returns a copy of the initial subgraph with the corresponding substitutions.*"``
  * ``{func}`aesara.clone_replace` ``

Also double check all the references in that notebook. The last sentence for example has a link to the writing a distribution guide which seems to point to the markdown file on github. Links to github don't trigger pre-commit to fail, but that should not be done. That should be a sphinx cross-reference.

Commit:
remove links and typos

juanitorduz commented:
Yeah! I was lazy skipping the `pre-commit` hooks while developing (my bad 🙈 ). In https://github.com/pymc-devs/pymc/pull/5721/commits/54841a3fd0c1b57714d0df6504fcc4d342220933 I believe I removed all links. Thanks!

OriolAbril commented:
> Yeah! I was lazy skipping the pre-commit hooks while developing (my bad see_no_evil ). In https://github.com/pymc-devs/pymc/commit/54841a3fd0c1b57714d0df6504fcc4d342220933 I believe I removed all links. Thanks!

Removing the links is not the solution. The problem is not with those links being there is with those links being in url format instead of sphinx cross-references

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1883 @@
+{
```
<p>new_w.name shoud be log of exp and not exp of log!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='42'/>

juanitorduz replied:
good catch! thanks!

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1883 @@
+{
```
<p>Perhaps print and mention the <code>x.owner.op</code> to show it is a RandomVariable. We introduced the term in the header but never referred to it afterwards. Also emphasize this is what is exactly the same that is created/returned when we call pm.Normal in the more familiar PyMC model below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

juanitorduz replied:
I tried to rephrase it following your suggestion. Feedback is welcome :)

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1883 @@
+{
```
<p>Emphasize that you are talking about the first input (rng) and also mention now you changed size as well. </p><p> </p><p> </p><p>Also a bit strange to first show pm.Normal.dist and then at.random.normal. Maybe the other way around? Or show both examples with at.random.normal and quickly mention you get the same object when you call pm.Normal.dist?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='55'/>

juanitorduz replied:
I tried to rephrase it following your suggestion. Feedback is welcome :)

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1883 @@
+{
```
<p>Not a shared variable, but a random variable</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='57'/>

juanitorduz replied:
+1

Commit:
Merge branch 'main' into docs/aesara_pymc_nb

Commit:
fix suggestions n + 1 XD

juanitorduz commented:
@OriolAbril I tried adding the links in https://github.com/pymc-devs/pymc/pull/5721/commits/fe4ca51c814197dd8d3a8142f331808428008e7a . I would like to verify the output on the build but I can not find it on https://pymcio--5721.org.readthedocs.build/projects/docs/en/5721/learn.html 🙈 . Any other suggestion with respect to style (I am not very familiar with sphinx).

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1960 @@
+{
```
<p>What's up with this warning?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='2'/>

juanitorduz replied:
I have no idea, I merged `main` into my developer branch.

juanitorduz replied:
maybe https://askubuntu.com/questions/1406952/what-is-the-meaning-of-this-pkgresourcesdeprecationwarning-warning-from-pipenv?

ricardoV94 replied:
<p>Seems to be gone! Is this something we need to fix in PyMC or was it specific to your installation? (I did not read the links yet)</p>

juanitorduz replied:
<p>It is gone because re-ran the cell twice. I believe it has to do with my installation as in the the PR preview is not there https://pymcio--5721.org.readthedocs.build/projects/docs/en/5721/learn/core_notebooks/pymc_aesara.html#introduction-to-aesara</p>

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1960 @@
+{
```
<p>Nitpick, we already said "One of the most interesting features of Aesara is..." in a previous paragraph. Perhaps "Another interesting feature of Aesara is"?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='35'/>

juanitorduz replied:
+1

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1960 @@
+{
```
<p>Small preference: I would show at.random.normal before pm.Normal.dist, just because we had been presenting Aesara operators until now.</p>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='53'/>

juanitorduz replied:
I added this suggestion in https://github.com/pymc-devs/pymc/pull/5721/commits/af34f1b473de3e7f652116b446bdb9f65f9853f9 :)

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1960 @@
+{
```
<p>Perhaps emphasize in text that we are just creating RandomVariables like we saw before, but now registering them in a PyMC model.</p><p> </p><p>And then it follows naturally to mention <code>model.basic_RVs</code> like you did </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

juanitorduz replied:
+1

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1960 @@
+{
```
<p>Nitpick, do you need the sum?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='91'/>

juanitorduz replied:
I think so otherwise we get a vector right?
![image](https://user-images.githubusercontent.com/22996444/172025045-aa7f233e-a045-45ea-a2cf-29a4bd1d29f3.png)

ricardoV94 replied:
<p>My bad, I got confused with the multivariate normal comparison! I assumed z was a multivariate normal as well</p>

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1960 @@
+{
```
<p>Avoid the word dispatch? I think there was more context related to dispatching before, but now it doesn't make as much sense</p><p>Maybe we don't need the cell at all? Or mention just logcdf, users won't have much needed for <code>moment</code> (it's not called <code>get_moment</code>  anymore).</p>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='92'/>

juanitorduz replied:
+1

ricardoV94 replied:
<p>Maybe "A similar function exists for <code>logcdf</code>"? The strategy part was related to the dispatching iirc  </p>

ricardoV94 replied:
<p>Should this have been changed?</p>

juanitorduz replied:
<p>good catch! I will correct this as your initial suggestion </p>

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,1960 @@
+{
```
<p>There's a missing <code>-</code> in <code>np.exp(10)</code> in the scipy call for the logpdf of x, that's why they don't match.</p><p> </p><p>Something is also odd with the sigma logp, it shouldn't match with the scipy because of the jacobian term (which should just be a -10)?</p>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='106'/>

juanitorduz replied:
Thanks! I added a comment about it. I do believe is the Jacobian.

ricardoV94 replied:
<p>Ah awesome, I missed the -10 at the beginning, I was looking at the end xD
</p><p> </p><p>World order is restored!</p>

Commit:
address further comments

juanitorduz commented:
Thank you @ricardoV94 for all of your valuable feedback! In https://github.com/pymc-devs/pymc/pull/5721/commits/af34f1b473de3e7f652116b446bdb9f65f9853f9 I try to address them all and left comments on the few which are open.

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>We can omit the manual creation of the rng variable (aesara will create one automatically we don't specify it). Or did you want to show it on purpose?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='53'/>

juanitorduz replied:
+1. I removed it.

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>Do we need this? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='71'/>

juanitorduz replied:
In a previous comment I thought you also wanted to compare the owners. I do believe is a bit redundant so I'll remove it.

juanitorduz replied:
<p>In a previous comment I thought you also wanted to compare the owners. I do believe is a bit redundant so will remove it.</p>

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>I think it's less confusing to call it "samples from a vector of two normal distributions", and label <code>z[0]</code> and <code>z[1]</code>. They are indeed equivalent to a specific mvnormal but I think that distracts from the goal here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='84'/>

juanitorduz replied:
<p>+1. Changed to "Samples Histogram"</p>

juanitorduz replied:
+1. Changed to "Samples Histogram"

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>Similarly here I would call <code>scipy.stats.normal.logpdf</code> and sum the output, as that is exactly what this PyMC example is doing.</p><p> </p><p>(Also then you don't need to sum here and below)</p>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='93'/>

juanitorduz replied:
<p>+1</p>

juanitorduz replied:
+1

ricardoV94 commented:
> Thank you @ricardoV94 for all of your valuable feedback! In https://github.com/pymc-devs/pymc/pull/5721/commits/af34f1b473de3e7f652116b446bdb9f65f9853f9 I try to address them all and left comments on the few which are open.

One more round and I think we are done on the content side. For the crosslink and formatting I will defer to @OriolAbril if he has the time.

A good and natural follow up to this notebook would be PyMC and Aeppl!

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>Perhaps mention after this block that Model also has methods to extract the gradient (dlogp) and hessian (d2logp) of the logp and link to the model entry in the docs?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='113'/>

juanitorduz replied:
<p>+1</p>

juanitorduz replied:
+1

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>"... internals of <code>aesara</code> RandomVariables and <code>pymc</code> distributions...”  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='114'/>

juanitorduz replied:
+1

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>I don't see any of the cross-references I suggested in <a href="https://github.com/pymc-devs/pymc/pull/5721#issuecomment-1146344919" target="_blank">this comment</a> (nor here nor in the other cases below)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

juanitorduz replied:
Thanks for your feedback @OriolAbril ! I think I managed to:
- [x] Include images
- [x] Include reference within `pymc` docs.

However, I have not managed to add the cross-references. For example:
- [ ] How to properly link to Aesara's doc? At the moment I am pointing to the GitHub repository.
- [ ] How to reference to a specific external method like `aesara.dprint`? 

I have done some research but got nowhere with these two open questions. Could you please give me some hints? Thanks!

OriolAbril replied:
Use the syntax I shared in https://github.com/pymc-devs/pymc/pull/5721#issuecomment-1146344919. For more context I'd recommend https://pymc-data-umbrella.xyz/en/latest/webinars/contributing_to_documentation/index.html#contributing-docs

juanitorduz replied:
Got it! this was indeed very helpful! I tried to add the cross-references to all methods and functions used in the notebook. I hope I did not miss anything!

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
The image is not rendering correctly: https://pymcio--5721.org.readthedocs.build/projects/docs/en/5721/learn/core_notebooks/pymc_aesara.html#introduction-to-aesara. Not sure how you added it to the notebook. If you aren't, I'd suggest using

<p> </p><pre class="ql-syntax" spellcheck="false">![aesara logo](https://raw.githubusercontent.com/aesara-devs/aesara/main/doc/images/aesara_logo_2400.png)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='4'/>

juanitorduz replied:
Thanks! These images came from the original notebook. Now that I know how to preview the notebook I will take care of the images 👍

juanitorduz replied:
<p>+1</p>

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>Image not rendering either. This should be added as an image file and commited to the core notebooks folder, then use markdown syntax to add the image with the alt text</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

juanitorduz replied:
I am getting it as the aesara logo via github.

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>It looks like this title is level 1 markdown title which feels a bit weird to me and makes the toc on the right sidebar look a bit strange. I think this should be at the same level as the "introduction to aesara" section. Also, this image isn't rendering correctly either.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='49'/>

juanitorduz replied:
<p>+1</p>

juanitorduz replied:
+1 on the markdown header. I will also fix the image.

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>Y label says frequency but the histogram is not normalized, I think it should be "counts" instead</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='51'/>

juanitorduz replied:
+1

OriolAbril commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2039 @@
+{
```
<p>This should also use the <code>ref</code> syntax, which to work will need you to edit that file and add the target at the top right before the title which seems to be missing. The target is the <code>(target)=</code> thing like you have added at the top of this file  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='114'/>

juanitorduz replied:
+1 Added!

Commit:
fix content feedback

Commit:
style corrections part 1

Commit:
attempt to adds cross-references

Commit:
fix methods references

Commit:
add some cross-references [WIP]

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2010 @@
+{
```
<p>I think here you can already remark that x is just a normal RandomVariables and that x and y are the same except for the rng. Then you don't need to do the comparison below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='61'/>

juanitorduz replied:
+1

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2010 @@
+{
```
<p>Let's use <code>model.compile_logp()(point)</code>?</p><p> </p><p><code>point_logps</code> is more of an internal specialized helper. <code>compile_logp</code> is more likely to be what users need and is more powerful.</p><p> </p><p>And, perhaps, before even using <code>compile_logp</code> we could call <code>model.logpt</code>,print the graph and emphasize this graph represents the model logpt but like before it does not do any computations until you compile and pass in the inputs. Then we coul use <code>.eval</code> for debugging, and then say one should compile it into a function for use in production, and that pymc model offers the helper <code>compile_logp</code> to make this easier.</p>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='88'/>

juanitorduz replied:
+1

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2010 @@
+{
```
<p>Should this cell have been removed?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='90'/>

juanitorduz replied:
you are totally right! I'll do so!

Commit:
more formating and final comments

juanitorduz commented:
Thanks for all your feedback! After addressing all comments (I hope) and a crash course on `sphinx`, it would be great if you can give me some final feedback. Thanks again.

ricardoV94 committed:
Present pm.logp before model utilities

ricardoV94 commented:
@juanitorduz I pushed a commit that shuffles a small portion of the guide around to present `pm.logp` before introducing the `Model.logpt` and `Model.compile_logp` methods. What do you think? Feel free to revert or modify if you don't like the changes. Also I wouldn't be surprised if I screw up the new sphinx references, so apologies in advance for that!

ricardoV94 commented:
Does the reviewnb not show difference per-commit or did I screw up something? Seems like the latter.

ricardoV94 committed:
Fix some references and minor changes

ricardoV94 committed:
Add Model reference

OriolAbril commented:
> Does the reviewnb not show difference per-commit

it does not. I skimmed the current preview and it looks good, we can continue to iterate and improve it further if needed. There are some references that are not working. I think they are wrong right now, but they also can't work. The root of the issue is #5282.

We are using `pm.logp` in the code, so the reference should be `` {func}`~pymc.logp` ``. However, that won't work because `pymc.logp` **is not documented anywhere**. It is fine to leave the references with the right syntax and not generating the correct links because once we document them they will be generated and point to the right place, it is not right to leave `pymc.distributions.logprob...` as references

OriolAbril commented:
Note, I was going to comment on the issue that the logp functions are missing but it is already there: https://github.com/pymc-devs/pymc/issues/5282#issuecomment-1068083637

OriolAbril commented:
Yes, **references should always be the same as what was used in the code** (with maybe adding the `~` before it so shorten the link text) and api pages should always document objects at that path. The file path where the object is implemented _is irrelevant to users_

ricardoV94 commented:
> Yes, **references should always be the same as what was used in the code** (with maybe adding the `~` before it so shorten the link text) and api pages should always document objects at that path. The file path where the object is implemented _is irrelevant to users_

I removed my comment, because I realized you had already answered :D

Commit:
correct func path references

juanitorduz commented:
Thanks for all the improvements 🚀 ! In https://github.com/pymc-devs/pymc/pull/5721/commits/79212ef9e52266bb6ad5d139b30a36955022f16f I changes the references of the functions which are not yet in the docs, but of course we want to have them right.

Anything else missing? (I might be slower to contribute during the working day so feel free to add / remove parts in the meantime)

ricardoV94 commented on docs/source/learn/core_notebooks/pymc_aesara.ipynb:
```diff
@@ -0,0 +1,2378 @@
+{
```
<p>"to obtain" instead of "to obtained"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5721/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='115'/>

juanitorduz replied:
Thx! Solved in https://github.com/pymc-devs/pymc/pull/5721/commits/4ce059a6fe34c9ca2aade794a8af36efbb875258

ricardoV94 commented:
Found a typo, otherwise looks 👌

Commit:
typo

twiecki commented:
This is awesome -- thanks @juanitorduz!

twiecki commented:
@OriolAbril how do we add this to the website?

ricardoV94 commented:
It's here: https://www.pymc.io/projects/docs/en/latest/learn/core_notebooks/pymc_aesara.html

One needs to switch the docs version to latest when coming from the main docs page

OriolAbril commented:
the notebooks from pymc examples and the gallery page gets updated automatically with every commit. But this notebook is not part of pymc-examples, it is part of the small subset of notebooks on core notebooks we keep here in the versioned docs. These are re-executed every time we build the docs and will always be in sync with our releases (a change in the library that breaks any of the code in these notebooks will break ci)
